test/provider_tracking.cpp

```diff
@@ -0,0 +1,370 @@
+        fprintf(stderr, "Alloc #%d\n", i);
+        ptr[i] = umfPoolAlignedMalloc(pools[i], size, utils_get_page_size());
+        ASSERT_NE(ptr[i], nullptr);
+
+        createPoolFromAllocation(ptr[i], size, &providers[i + 1],
+                                 &pools[i + 1]);
+    }
+
+    int s = TEST_LEVEL_SUCCESS;
+    fprintf(stderr, "Alloc #%d\n", s);
+    ptr[s] = umfPoolAlignedMalloc(pools[s], size, utils_get_page_size());
+    ASSERT_NE(ptr[s], nullptr);
+
+    fprintf(stderr, "Free #%d\n", s);
+    umf_result = umfPoolFree(pools[s], ptr[s]);
+    ASSERT_EQ(umf_result, UMF_RESULT_SUCCESS);
+
+    for (int i = TEST_LEVEL_SUCCESS - 1; i >= 0; i--) {
+        umfPoolDestroy(pools[i + 1]);
+        umfMemoryProviderDestroy(providers[i + 1]);
+
+        fprintf(stderr, "Free #%d\n", i);
+        umf_result = umfPoolFree(pools[i], ptr[i]);
+        ASSERT_EQ(umf_result, UMF_RESULT_SUCCESS);
+    }
+}
+
+TEST_P(TrackingProviderTest, failure_exceeding_levels) {
+    umf_result_t umf_result;
+    size_t size;
+    void *ptr[MAX_ARRAY] = {0};
+    umf_memory_provider_handle_t providers[MAX_ARRAY] = {0};
+    umf_memory_pool_handle_t pools[MAX_ARRAY] = {0};
+
+    size = FIXED_BUFFER_SIZE - (2 * page_size);
+    pools[0] = pool.get();
+
+    for (int i = 0; i < TEST_LEVEL_FAILURE; i++) {
+        fprintf(stderr, "Alloc #%d\n", i);
+        ptr[i] = umfPoolAlignedMalloc(pools[i], size, utils_get_page_size());
+        ASSERT_NE(ptr[i], nullptr);
+
+        createPoolFromAllocation(ptr[i], size, &providers[i + 1],
+                                 &pools[i + 1]);
+    }
+
+    // tracker level is too high
+    int f = TEST_LEVEL_FAILURE;
+    fprintf(stderr, "Alloc #%d\n", f);
+    ptr[f] = umfPoolAlignedMalloc(pools[f], size, utils_get_page_size());
+    ASSERT_EQ(ptr[f], nullptr);
+
+    for (int i = TEST_LEVEL_FAILURE - 1; i >= 0; i--) {
+        umfPoolDestroy(pools[i + 1]);
+        umfMemoryProviderDestroy(providers[i + 1]);
+
+        fprintf(stderr, "Free #%d\n", i);
+        umf_result = umfPoolFree(pools[i], ptr[i]);
+        ASSERT_EQ(umf_result, UMF_RESULT_SUCCESS);
+    }
+}
+
+TEST_P(TrackingProviderTest, reverted_free_half_size) {
+    umf_result_t umf_result;
+    size_t size0;
+    size_t size1;
+    void *ptr0 = nullptr;
+    void *ptr1 = nullptr;
+
+    umf_memory_pool_handle_t pool0 = pool.get();
+
+    size0 = FIXED_BUFFER_SIZE - (2 * page_size);
+    ptr0 = umfPoolAlignedMalloc(pool0, size0, utils_get_page_size());
+    ASSERT_NE(ptr0, nullptr);
+
+    umf_memory_provider_handle_t provider1 = nullptr;
+    umf_memory_pool_handle_t pool1 = nullptr;
+    createPoolFromAllocation(ptr0, size0, &provider1, &pool1);
+
+    size1 = size0 / 2; // half size
+
+    ptr1 = umfPoolMalloc(pool1, size1);
+    ASSERT_NE(ptr1, nullptr);
+
+    // try to free the pointer from the first pool (half size)
+    umf_result = umfPoolFree(pool0, ptr0);
+    ASSERT_EQ(umf_result, UMF_RESULT_ERROR_INVALID_ARGUMENT);
```

Comment:
Hmm, that is interesting. What should be the behavior in that case? The code is definitely ill-formed from the user's perspective, but the question is how we should handle it. Formally the argument is valid: `ptr0` was allocated from the `pool0` it is not clear if we should return error code, because on the L331 UMF return success for the same arguments.

Reply:
It has to fail here, since `umfMemoryProviderFree()` fails because of a wrong size. The only question is if we can return a different error here.

Reply:
@vinser52 Now https://github.com/oneapi-src/unified-memory-framework/pull/1157 is included in this PR and UMF does NOT return success for the same arguments at the L331.